src/groups/mqb/mqbc/mqbc_clusterutil.cpp

```diff
@@ -970,30 +966,56 @@ ClusterUtil::assignQueue(ClusterState*         clusterState,
-            local::panic(domIt->second->domain());
-        }
-        else {
-            const int watermark = static_cast<int>(
-                maxQueues * k_MAX_QUEUES_HIGH_WATERMARK);
-            if (registeredQueues < watermark && requestedQueues >= watermark) {
-                local::alarm(domIt->second->domain(), requestedQueues);
+    if (queueIt == domIt->second->queuesInfo().end()) {
+        BSLS_ASSERT_SAFE(previousState ==
+                         ClusterStateQueueInfo::State::k_NONE);
+
+        // Need to check if we have capacity before we allocate resources
+        // for this new queue.  The current number of registered queues is:
+        // num(assigned) + num(assigning) + num(unassigning).
+        const int registeredQueues = static_cast<int>(
+            domIt->second->queuesInfo().size());
+        const int maxQueues = domIt->second->domain()->config().maxQueues();
+        if (maxQueues != 0) {
+            const int requestedQueues = registeredQueues + 1;
+            if (requestedQueues > maxQueues) {
+                local::panic(domIt->second->domain());
             }
-        }
-
-        if (requestedQueues > maxQueues) {
-            if (status) {
-                status->category() = bmqp_ctrlmsg::StatusCategory::E_REFUSED;
-                status->code()     = mqbi::ClusterErrorCode::e_LIMIT;
-                status->message()  = k_MAXIMUM_NUMBER_OF_QUEUES_REACHED;
+            else {
+                const int watermark = static_cast<int>(
+                    maxQueues * k_MAX_QUEUES_HIGH_WATERMARK);
+                if (registeredQueues < watermark &&
+                    requestedQueues >= watermark) {
+                    local::alarm(domIt->second->domain(), requestedQueues);
+                }
             }
 
-            return QueueAssignmentResult::k_ASSIGNMENT_REJECTED;  // RETURN
+            if (requestedQueues > maxQueues) {
+                if (status) {
+                    status->category() =
+                        bmqp_ctrlmsg::StatusCategory::E_REFUSED;
+                    status->code()    = mqbi::ClusterErrorCode::e_LIMIT;
+                    status->message() = k_MAXIMUM_NUMBER_OF_QUEUES_REACHED;
+                }
+
+                return QueueAssignmentResult::k_ASSIGNMENT_REJECTED;  // RETURN
+            }
         }
+
+        // We have capacity and can add this queue to the collection.
+        // The queue will be in k_ASSIGNING state until we commit queue
+        // assignment advisory.
+        QueueInfoSp queueInfo;
+
+        queueInfo.createInplace(allocator, uri, allocator);
+
+        queueIt = domIt->second->queuesInfo().emplace(uri, queueInfo).first;
+    }
+    else {
+        // Note that we already have `queueIt` and allocated resources for this
+        // queue.  No need to allocate new QueueInfo and check capacity.
+        BSLS_ASSERT_SAFE(previousState ==
+                         ClusterStateQueueInfo::State::k_UNASSIGNING);
```

Comment:
Note that we will cannot have an alarm if we transit `k_UNASSIGNING` -> `k_ASSIGNING`, because we already have allocated resources and valid iterator `queueIt`